In [12]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from random import randint
import math
import warnings
warnings.filterwarnings("ignore")

In [13]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [14]:
final_training_df = pd.read_csv(r'/content/drive/MyDrive/Research Paper/DATA/final_training_df.csv')
final_testing_df = pd.read_csv(r'/content/drive/MyDrive/Research Paper/DATA/final_testing_df.csv')

print(final_training_df.shape)
print(final_testing_df.shape)
print("The Datasets have been loaded...")

(250, 11)
(50, 11)
The Datasets have been loaded...


In [15]:
X_train = final_training_df.drop(['condition'], axis = 1)
Y_train = final_training_df['condition']

Y_train = np.array(Y_train)

X_test = final_testing_df.drop(['condition'], axis = 1)
Y_test = final_testing_df['condition']

Y_test = np.array(Y_test)

print(X_train.shape, Y_train.shape)
print(X_test.shape, Y_test.shape)

(250, 10) (250,)
(50, 10) (50,)


In [16]:
X_train.head()

Mean         SD  Kurtosis  ...     range       RMS  Shape factor
0 -1.166491 -0.641888 -0.631950  ... -0.740012  0.033720     -0.413015
1  0.389329  0.830822  1.604929  ...  1.196487  0.744678     -0.432032
2  1.439655 -0.488640  0.576646  ...  0.236606  0.108506     -0.619835
3 -0.641116 -0.718663 -0.972769  ... -1.093864 -0.003179     -0.486737
4  0.658361 -0.764562 -0.388668  ... -0.008617 -0.024884     -0.598013

[5 rows x 10 columns]

In [17]:
Y_train

array([5, 3, 1, 5, 1, 1, 1, 1, 0, 3, 3, 1, 1, 4, 3, 3, 4, 0, 1, 4, 5, 3,
       0, 4, 1, 3, 2, 0, 2, 0, 3, 5, 5, 1, 2, 3, 0, 4, 5, 2, 1, 4, 1, 3,
       0, 3, 5, 4, 4, 1, 5, 4, 2, 4, 3, 5, 4, 5, 3, 0, 2, 2, 4, 1, 5, 3,
       2, 5, 1, 2, 3, 1, 0, 3, 0, 4, 2, 2, 2, 1, 2, 4, 3, 1, 0, 3, 3, 4,
       5, 4, 1, 5, 5, 1, 2, 1, 5, 3, 4, 2, 5, 2, 0, 2, 5, 0, 0, 0, 0, 1,
       2, 3, 2, 5, 5, 3, 2, 3, 4, 0, 5, 2, 4, 4, 0, 1, 0, 1, 5, 1, 2, 4,
       5, 5, 2, 0, 2, 3, 1, 4, 2, 4, 0, 1, 5, 4, 3, 4, 2, 1, 5, 4, 5, 1,
       2, 2, 3, 5, 3, 3, 5, 1, 0, 0, 1, 3, 1, 0, 3, 3, 1, 5, 0, 4, 0, 1,
       3, 4, 4, 5, 3, 3, 3, 3, 2, 4, 0, 2, 5, 4, 2, 4, 1, 1, 0, 0, 3, 1,
       3, 3, 4, 2, 4, 5, 2, 4, 5, 2, 0, 0, 2, 2, 0, 2, 1, 5, 4, 0, 1, 5,
       0, 0, 3, 1, 1, 4, 5, 4, 2, 0, 2, 0, 0, 5, 5, 0, 0, 4, 1, 2, 5, 0,
       3, 5, 0, 2, 5, 2, 4, 4])

# TO DEFINE A GENERAL FUNCTION THAT WOULD GIVE A 10-FOLD CV ACCURACY FOR EACH KERNEL

In [ ]:
def to_check_accuracy_of_each_kernel_on_cv(X, Y, kernel_type):

  from sklearn.svm import SVC
  from sklearn.metrics import accuracy_score
  from sklearn.model_selection import KFold

  # to create a list which wil store the accuracy of all the elephants i.e. SVM parameters
  accuracy_list = []

  # to reset the index of the data which is passed to the function
  X = X.reset_index(drop=True)


  '''
  so now, my training data which is of the size (250,10) will be split into 10 folds, where everytime it will be validated on each fold
  and trained on the remaining 9 folds. 
  ''' 
  kf = KFold(n_splits = 10, random_state = None, shuffle = False)


  total_score = 0

  for train_index, test_index in kf.split(X):

    X_train_fold, X_cv_fold = X.iloc[train_index], X.iloc[test_index]
    Y_train_fold, Y_cv_fold = Y[train_index], Y[test_index]

    # now to use this data to train the model and store the validation accuracy for every fold
    svm_model = SVC(kernel=kernel_type, verbose = False)
    svm_model.fit(X_train_fold, Y_train_fold)

    predicted_value = svm_model.predict(X_cv_fold)
    score = accuracy_score(Y_cv_fold, predicted_value)

    total_score = total_score + score

  final_accuracy = total_score/10

  return final_accuracy

In [ ]:
def to_check_accuracy_on_training_and_testing_data(X_train, Y_train, X_test, Y_test, kernel_type):

  from sklearn.svm import SVC
  from sklearn.metrics import accuracy_score

  clf = SVC(kernel = kernel_type)
  clf.fit(X_train, Y_train)

  test_predicted_output = clf.predict(X_test)
  train_predicted_output = clf.predict(X_train)

  testing_accuracy = accuracy_score(Y_test, test_predicted_output)
  training_accuracy = accuracy_score(Y_train, train_predicted_output)

  return training_accuracy, testing_accuracy

# RBF KERNEL

In [ ]:
training_accuracy, testing_accuracy = to_check_accuracy_on_training_and_testing_data(X_train, Y_train, X_test, Y_test, "rbf")

print("TRAINING ACCURACY:- ", training_accuracy)
print("TESTING ACCURACY:- ", testing_accuracy)

TRAINING ACCURACY:-  0.956
TESTING ACCURACY:-  0.9


In [ ]:
'''
To check the accuracy of RBF Kernel with 10 fold CV
'''
rbf_kernel = to_check_accuracy_of_each_kernel_on_cv(X_train, Y_train, "rbf")

print("The 10 Fold-CV accuracy on RBF Kernel is:- ", rbf_kernel)

The 10 Fold-CV accuracy on RBF Kernel is:-  0.932


# LINEAR KERNEL

In [ ]:
training_accuracy, testing_accuracy = to_check_accuracy_on_training_and_testing_data(X_train, Y_train, X_test, Y_test, "linear")

print("TRAINING ACCURACY:- ", training_accuracy)
print("TESTING ACCURACY:- ", testing_accuracy)

TRAINING ACCURACY:-  0.964
TESTING ACCURACY:-  0.88


In [ ]:
'''
To check the accuracy of Linear Kernel with 10 fold CV
'''
linear_kernel = to_check_accuracy_of_each_kernel_on_cv(X_train, Y_train, "linear")

print("The 10 Fold-CV accuracy on RBF Kernel is:- ", linear_kernel)

The 10 Fold-CV accuracy on RBF Kernel is:-  0.9359999999999999


# SIGMOID KERNEL

In [ ]:
training_accuracy, testing_accuracy = to_check_accuracy_on_training_and_testing_data(X_train, Y_train, X_test, Y_test, "sigmoid")

print("TRAINING ACCURACY:- ", training_accuracy)
print("TESTING ACCURACY:- ", testing_accuracy)

TRAINING ACCURACY:-  0.82
TESTING ACCURACY:-  0.76


In [ ]:
'''
To check the accuracy of Linear Kernel with 10 fold CV
'''
sigmoid_kernel = to_check_accuracy_of_each_kernel_on_cv(X_train, Y_train, "sigmoid")

print("The 10 Fold-CV accuracy on RBF Kernel is:- ", sigmoid_kernel)

The 10 Fold-CV accuracy on RBF Kernel is:-  0.812


# POLYNOMIAL KERNEL

In [ ]:
training_accuracy, testing_accuracy = to_check_accuracy_on_training_and_testing_data(X_train, Y_train, X_test, Y_test, "poly")

print("TRAINING ACCURACY:- ", training_accuracy)
print("TESTING ACCURACY:- ", testing_accuracy)

TRAINING ACCURACY:-  0.86
TESTING ACCURACY:-  0.84


In [ ]:
'''
To check the accuracy of Linear Kernel with 10 fold CV
'''
polynomial_kernel = to_check_accuracy_of_each_kernel_on_cv(X_train, Y_train, "poly")

print("The 10 Fold-CV accuracy on RBF Kernel is:- ", polynomial_kernel)

The 10 Fold-CV accuracy on RBF Kernel is:-  0.8399999999999999


# TO CHECK THE ACCURACY ON GRID SEARCH CV

In [27]:
'''
as it is already proved that RBF kernel is the most suitable one, we are using the same one
'''
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC

model_svc_gridsearch = SVC(kernel = 'rbf')

regularization_parameter = [1, 5, 15, 30, 45, 60, 75, 90]
gamma = [1, 4, 8, 12, 16, 25, 40, 50]

param_dict = dict(C = regularization_parameter, gamma = gamma)

clf_gridsearchcv = GridSearchCV(estimator = model_svc_gridsearch, param_grid = param_dict, cv=10, scoring='f1_micro', verbose=True,
                              n_jobs = -1, return_train_score=True)

clf_gridsearchcv.fit(X_train, Y_train)

Fitting 10 folds for each of 64 candidates, totalling 640 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done 216 tasks      | elapsed:    3.4s
[Parallel(n_jobs=-1)]: Done 640 out of 640 | elapsed:    7.8s finished


GridSearchCV(cv=10, error_score=nan,
             estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                           class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='scale', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='deprecated', n_jobs=-1,
             param_grid={'C': [1, 5, 15, 30, 45, 60, 75, 90],
                         'gamma': [1, 4, 8, 12, 16, 25, 40, 50]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
             scoring='f1_micro', verbose=True)

In [28]:
best_parameters = clf_gridsearchcv.best_params_
best_score = clf_gridsearchcv.best_score_

print("The values for the best hyperparameters for GRID SEARCH CV are:- ", best_parameters)
print("="*100)
print("The best CV score for GRID SEARCH CV is:- ", best_score)

The values for the best hyperparameters for GRID SEARCH CV are:-  {'C': 1, 'gamma': 1}
The best CV score for GRID SEARCH CV is:-  0.9200000000000002


In [29]:
results_gridsearchcv = pd.DataFrame.from_dict(clf_gridsearchcv.cv_results_)
results_gridsearchcv

mean_fit_time  std_fit_time  ...  mean_train_score  std_train_score
0        0.010769      0.001196  ...          0.982667         0.006425
1        0.011876      0.002154  ...          1.000000         0.000000
2        0.010527      0.000316  ...          1.000000         0.000000
3        0.011434      0.002052  ...          1.000000         0.000000
4        0.011403      0.002032  ...          1.000000         0.000000
..            ...           ...  ...               ...              ...
59       0.010843      0.001654  ...          1.000000         0.000000
60       0.011203      0.001670  ...          1.000000         0.000000
61       0.011293      0.001142  ...          1.000000         0.000000
62       0.010961      0.000712  ...          1.000000         0.000000
63       0.010467      0.000612  ...          1.000000         0.000000

[64 rows x 32 columns]

**Best 10 Fold-CV accuracy using gird search cv is 92% and the optimum parameters are C=1 and gamma=1.** 

I have used Micro F1 score instead of Accuracy as both of them give the same results as our's is not a MULTILABEL classification problem. It is not possible to use Accuracy as the metric in GridSearch CV for multiclass classification problem**

# RANDOM SEARCH CV

In [30]:
regularization_parameter = np.random.uniform(low = 0.1, high = 100, size = 8)
gamma = np.random.uniform(low = 0.1, high = 30, size = 8)

print(type(gamma))

<class 'numpy.ndarray'>


In [31]:
'''
as it is already proved that RBF kernel is the most suitable one, we are using the same one. To use uniform random distribution 
for initializing 8 parameters each
'''
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC

model_svc_randomsearch = SVC(kernel = 'rbf')

regularization_parameter = np.random.uniform(low = 0.1, high = 100, size = 8)
gamma = np.random.uniform(low = 0.1, high = 30, size = 8)

param_dict = dict(C = regularization_parameter, gamma = gamma)

clf_randomsearchcv = GridSearchCV(estimator = model_svc_randomsearch, param_grid = param_dict, cv=10, scoring='f1_micro', verbose=True,
                              n_jobs = -1, return_train_score=True)

clf_randomsearchcv.fit(X_train, Y_train)

Fitting 10 folds for each of 64 candidates, totalling 640 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done 216 tasks      | elapsed:    3.5s
[Parallel(n_jobs=-1)]: Done 640 out of 640 | elapsed:    7.9s finished


GridSearchCV(cv=10, error_score=nan,
             estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                           class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='scale', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='deprecated', n_jobs=-1,
             param_grid={'C': array([95.3097052 , 28.39447233, 29.06587542, 98.86495763, 64.36879245,
       50.15953336, 90.08793143, 31.36289207]),
                         'gamma': array([15.36738659, 14.86822961, 12.390538  , 26.09618815, 27.27630937,
       29.5280132 ,  5.96711701, 27.55862768])},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
             scoring='f1_micro', verbose=True)

In [32]:
best_parameters = clf_randomsearchcv.best_params_
best_score = clf_randomsearchcv.best_score_

print("The values for the best hyperparameters for RANDOM SEARCH CV are:- ", best_parameters)
print("="*100)
print("The best CV score for RANDOM SEARCH CV is:- ", best_score)

The values for the best hyperparameters for RANDOM SEARCH CV are:-  {'C': 95.30970519554722, 'gamma': 5.967117014667493}
The best CV score for RANDOM SEARCH CV is:-  0.7


In [33]:
results_randomsearchcv = pd.DataFrame.from_dict(clf_randomsearchcv.cv_results_)
results_randomsearchcv

mean_fit_time  std_fit_time  ...  mean_train_score  std_train_score
0        0.012071      0.001267  ...               1.0              0.0
1        0.012224      0.001938  ...               1.0              0.0
2        0.010513      0.000451  ...               1.0              0.0
3        0.011173      0.000826  ...               1.0              0.0
4        0.011522      0.001104  ...               1.0              0.0
..            ...           ...  ...               ...              ...
59       0.010785      0.000271  ...               1.0              0.0
60       0.011929      0.003658  ...               1.0              0.0
61       0.010898      0.000921  ...               1.0              0.0
62       0.010286      0.000098  ...               1.0              0.0
63       0.011203      0.001754  ...               1.0              0.0

[64 rows x 32 columns]

**Best 10 Fold-CV accuracy using random search cv is 70% and the optimum parameters are C=95.30970519554722 and gamma=5.967117014667493.** 

I have used Micro F1 score instead of Accuracy as both of them give the same results as our's is not a MULTILABEL classification problem. It is not possible to use Accuracy as the metric in GridSearch CV for multiclass classification problem**

# SIMPLE SVM (WITH LINEAR KERNEL)

The 10 Fold CV accuracy here is 93.59 with the dafault parameters used in sklearn i.e. C = 1 and gamma = 'scale'